### DISCUSSION TITLE
How to get the text linked to layers?

### DISCUSSION THREAD

**Discussion by jase64** [pymupdf/PyMuPDF#2301] [Q&A] [answered]
Hi Community,

First of all, thanks a lot to the developers (and other helpers) to bring that great library for struggling against the PDF format (mess?).

I currently have a set of PDFs that were generated by Autocad (version 14). The PDFs have been constructed taken into consideration the layers defined in Autocad.  I'm trying to read the text "contained" in (or controlled by) those layers.
I first tried a basic text extraction
```python
import fitz
doc = fitz.open(filename)
page = doc[0]
text = page.get_text("text")
```
Unfortunately, `text`does not return the expected data.
So, I tried to retrieve the ocgs
```python
ocgs = doc.get_ocgs()
```
'ocgs' returns a dictionary with layer ids and names
```python
{...
32: {'name': 'PROCESS', 'intent': [], 'on': True, 'usage': None},
 12: {'name': 'EQUIPMENT', 'intent': [], 'on': True, 'usage': None},
 38: {'name': 'VENDOR', 'intent': [], 'on': True, 'usage': None},
...}
```
Then I tried to dig into the `xrefs`
```python
len_xref=doc.xref_length()
for xref in range(1, len_xref):
    print('')
    print(f"Object {xref}, stream: {doc.xref_is_stream(xref)}")
    print(doc.xref_object(xref, compressed=False))
```
It returns again a list of objects, but no link to the actual content of the objects.
```
...
Object 12, stream: False
<<
  /Name (EQUIPMENT)
  /Type /OCG
>>
...
```
Has anyone an idea how to get access to the text contained in layers? in the oc? If so, how to link the layer to the 

PS: for the time being I don't want to use OCR extraction. Only if no other choice.

**Comment by JorjMcKie** [comment, marked as answer]
Thanks for the nice feedback!

Before a text extraction, you could temporarily switch to the desired layer using `doc.set_layer_ui_config(number, action=0)`. This is what you would do using a supporting PDF viewer. Available configurations can be display by `doc.layer_ui_configs()`.
The *action* parameter is 0 = set on (default), 1 = toggle on/off, 2 = set off.

**Comment by jase64** [comment]
Hi Jorj,
Thanks a lot for the hint. It worked fine. I didn't realize it could as simple as turning off all layers except the one of interest and then make a simple `page.get_text('text')`. 
Great work!
